src/geometry/euclidean/congruence.lean

```diff
@@ -0,0 +1,303 @@
+by {ext x, fin_cases x; refl}
+lemma _root_.fin.fin_3_swap_right {α : Type*} (a b c : α) : ![a,b,c] = ![a,c,b] ∘ ![0,2,1] :=
+by {ext x, fin_cases x; refl}
+
+
+lemma triangle_swap_left (h : ![a,b,c] ≅ ![d,e,f]) : ![b,a,c] ≅ ![e,d,f] :=
+begin
+  rw [fin.fin_3_swap_left b a c, fin.fin_3_swap_left e d f],
+  apply sub_congruence ![1,0,2] h,
+end
+lemma triangle_swap_right (h : ![a,b,c] ≅ ![d,e,f]) : ![a,c,b] ≅ ![d,f,e] :=
+begin
+  rw [fin.fin_3_swap_right a c b, fin.fin_3_swap_right d f e],
+  apply sub_congruence ![0,2,1] h,
+end
+
+lemma side_side_side {a b c : P₁} {d e f : P₂} (hs₁ : dist a b = dist d e)
+  (hs₂ : dist a c = dist d f) (hs₃ : dist b c = dist e f) : ![a,b,c] ≅ ![d,e,f] :=
+begin
+  intro i₁, fin_cases i₁; intro i₂; fin_cases i₂;
+    simp [dist_comm]; assumption,
+end
+
+include V₁
+
+lemma cos_angle_elim : real.cos (∠ a b c) =
+  (dist a b * dist a b + dist b c * dist b c - dist a c * dist a c) / (2 * dist a b * dist b c) :=
+begin
+  unfold euclidean_geometry.angle, rw inner_product_geometry.cos_angle,
+  rw real_inner_eq_norm_mul_self_add_norm_mul_self_sub_norm_sub_mul_self_div_two,
+  simp [ ← normed_add_torsor.dist_eq_norm'],
+  rw [div_div, mul_assoc, dist_comm b c],
+end
+
+
+
+lemma sin_angle_elim : real.sin (∠ a b c) =
+  real.sqrt (1 - ((dist a b ^ 2 + dist b c ^ 2 - dist a c ^ 2) ^ 2) /
+    (2 * dist a b * dist b c) ^ 2) :=
+begin
+  rw real.sin_eq_sqrt_one_sub_cos_sq
+    (euclidean_geometry.angle_nonneg a b c)
+    (euclidean_geometry.angle_le_pi a b c),
+  rw [cos_angle_elim, ← pow_two, ← pow_two, ← pow_two, div_pow],
+end
+
+omit V₁
+
+/-
+the library contains a similar lemma, but it requires the numerator to be positive instead of
+the denominator, so this is more useful in this case.
+-/
+lemma _root_.real.sqrt_div_sq :
+  ∀ {y : ℝ}, 0 ≤ y → ∀ (x : ℝ), real.sqrt (x / y^2) = real.sqrt x / y :=
+begin
+  intros y hy x,
+  rw [div_eq_inv_mul, div_eq_inv_mul],
+  rw [real.sqrt_mul, real.sqrt_inv, real.sqrt_sq hy],
+  rw inv_nonneg, exact sq_nonneg y,
+end
+
+
+
+include V₁ V₂
+
+lemma side_angle_side (ha : ∠ a b c = ∠ d e f)
+  (hs₁ : dist a b = dist d e) (hs₂ : dist b c = dist e f) : ![a,b,c] ≅ ![d,e,f] :=
+begin
+  refine side_side_side hs₁ _ hs₂,
+  by_cases h : dist a b = 0,
+  { rw h at hs₁,
+    replace hs₁ := hs₁.symm,
+    rw dist_eq_zero at hs₁ h,
+    rw [h, hs₁], exact hs₂ },
+  by_cases h' : dist b c = 0,
+  { rw h' at hs₂,
+    replace hs₂ := hs₂.symm,
+    rw dist_eq_zero at hs₂ h',
+    rw [← h', ← hs₂], exact hs₁ },
+
+  apply_fun real.cos at ha,
+  simp [cos_angle_elim, ← hs₁, ← hs₂] at ha,
+  rw div_left_inj'
+    (mul_ne_zero_iff.2 ⟨mul_ne_zero_iff.2 ⟨show (2:ℝ) ≠ 0, from by simp, h⟩, h'⟩) at ha,
+  rw sub_right_inj at ha,
+  rw [← pow_two , ← pow_two] at ha,
+  rw sq_eq_sq dist_nonneg dist_nonneg at ha,
+  exact ha,
+
+end
+
+omit V₂
+
+lemma sine_rule (hac : a ≠ c) (hbc : b ≠ c) :
+  real.sin (∠ a b c) / dist a c = real.sin (∠ b a c) / dist b c :=
```

Comment:
This should have a doc string pointing to the versions of the sine rule (involving the circumradius) in `geometry.euclidean.angle.sphere`. (It's still appropriate to prove the version without the circumradius separately - though I think it should go in `geometry.euclidean.triangle` for now - because of lower dependencies; it's more likely that `geometry.euclidean.angle.sphere` will have a use for importing congruences in future than the other way round.)